src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Finitely_critical_filtrations.h

```diff
@@ -0,0 +1,1051 @@
+      stuff.push_to(x);
+    }
+  }
+  // TODO : this is not well defined for kcritical <-> kcritical
+
+  /** \brief This functions take the filtration value `this` and pulls it to the
+   * cone \f$ \{ y\in \mathbb R^n : y<=x \} \f$. After calling this method, the
+   * value of this is updated to \f$ \mathrm{this} = \max \{ y\in \mathbb R^n :
+   * y<=this \}\cap \{ y\in \mathbb R^n : y<=x \}
+   * @param[in] x The target filtration value on which to push `this`.
+   */
+  inline void pull_to(const One_critical_filtration<T> &x) {
+    if (x.is_inf() || this->is_nan() || x.is_nan() || this->is_minus_inf())
+      return;
+    if (this->is_inf() || x.is_minus_inf()) {
+      *this = x;
+      return;
+    }
+    for (auto &stuff : *this) {
+      stuff.pull_to(x);
+    }
+  } 
+  // cannot be const, as gudhi iterators are not const
+  template <typename U> inline U linear_projection(const std::vector<U> &x) {
+    if constexpr (co) {
+      U projection = std::numeric_limits<U>::lowest();
+      for (const auto &y : *this) {
+        projection = std::max(projection, y.linear_projection(x));
+      }
+      return projection;
+    } else {
+      U projection = std::numeric_limits<U>::max();
+      for (auto &y : *this) { // cannot be const (Gudhi)
+        projection = std::min(projection, y.linear_projection(x));
+      }
+      return projection;
+    }
+  }
+
+  /*
+  * Checks if b is cleanable with respect to a
+  */
+  static inline bool dominates(const OneCritical&a, const OneCritical&b, value_type max_error) {
+    if constexpr (co)
+      return a - max_error <= b;
+    else {
+      return a + max_error >= b;
+    }
+  }
+
+  static inline bool dominates(const OneCritical&a, const OneCritical&b) {
+    if constexpr (co)
+      return a <= b;
+    else {
+      return a >= b;
+    }
+  }
+  static inline bool strictly_dominates(const OneCritical&a, const OneCritical&b) {
+    if constexpr (co)
+      return a < b;
+    else {
+      return a > b;
+    }
+  }
+
+  /*
+  * Same method as the one in OneCriticalFiltration. 
+  * Given a grid, and assuming that `this` are the coordianates
+  * in this grid, evaluate `this` into this grid
+  */
+  template <typename U>
+  inline Multi_critical_filtration<U>
+  evaluate_in_grid(const std::vector<std::vector<U>>& grid) const {
+    Multi_critical_filtration<U> out(this->num_generators());
+    for (std::size_t i = 0; i < this->num_generators(); ++i){
+      out[i] = this->operator[](i).evaluate_in_grid(grid);
+    }
+    return out;
+  }
+  
+  /*
+  * Remove redundant values 
+  */
+  inline void clean(value_type max_error = 0) {
+    // A bit ugly, todo : erase+removeif ?
+    for (std::size_t i = 0; i < multi_filtration_.size(); i++) {
+      for (std::size_t j = 0; j < multi_filtration_.size(); j++) {
+        if (i == j)
+          continue;
+        if (dominates(multi_filtration_[j], multi_filtration_[i], max_error)) {
+          multi_filtration_[i].clear();
```

Comment:
Are future calls to `dominates` happy to encounter an empty vector? Or does that somehow not happen?